src/lib/ui/index.js

```diff
@@ -0,0 +1,90 @@
+const os = require('os');
+const Chalk = require('chalk').constructor;
+const platformsById = require('../../cmd/constants').platformsById;
+
+
+module.exports = class UI {
+	constructor({
+		stdin = process.stdin,
+		stdout = process.stdout,
+		stderr = process.stderr
+	} = {}){
+		this.stdin = stdin;
+		this.stdout = stdout;
+		this.stderr = stderr;
+		this.chalk = new Chalk(); // TODO (mirande): explicitly enable / disable colors
+		this.EOL = os.EOL;
+	}
+
+	write(data){
+		const { stdout, EOL } = this;
+		stdout.write(data + EOL);
+	}
+
+	error(data){
+		const { stderr, EOL } = this;
+		stderr.write(data + EOL);
+	}
+
+	logDeviceDetail(devices){
+		const { EOL, chalk } = this;
+		const deviceList = Array.isArray(devices) ? devices : [devices];
+		const lines = [];
+
+		for (let i = 0; i < deviceList.length; i++){
+			const device = deviceList[i];
+			const deviceType = platformsById[device.product_id] || `Product ${device.product_id}`;
+			const connectedState = device.connected ? 'online' : 'offline';
+			let name;
+
+			if (!device.name || device.name === 'null'){
+				name = '<no name>';
+			} else {
+				name = device.name;
+			}
+
+			if (device.connected){
```

Comment:
Regretably the user device API endpoint returns `device.connected` and the product device API endpoint returns `device.online` when a device is online :cry: 
So you have to check both to report if a device is online.

Reply:
fixed over here 👉 https://github.com/particle-iot/particle-cli/pull/544/commits/9006100a68a1991e9a1def1fd509995631e09080